Add method to emitting env placeholders

davidfowl opened the pull request:
- AsEnvironmentPlaceholder will add an environment variable place holder for docker service/file values.
- Unknown parameters become env placeholders

## Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] Yes
- Does the change make any security assumptions or guarantees?
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
  - [x] No

davidfowl committed:
Add method to emitting env placeholders
- AsEnvPlaceHolder will add an envrionment variable place holder for docker service/file values.
- Unknown parameters become env placeholders

mitchdenny commented on tests/Aspire.Hosting.Docker.Tests/DockerComposePublisherTests.cs:
```diff
@@ -160,18 +162,24 @@ public async Task DockerComposeAppliesServiceCustomizations()
                 // Set a restart policy
                 composeService.Restart = "always";
 
+                composeService.ContainerName = containerNameParam.AsEnvPlaceHolder(serviceResource);
```
I'm not sure about the extension name here. Maybe `AsEnvironmentPlaceholder` (note Env => Environment and capitalization of Placeholder).

mitchdenny replied:
Mark this API as experimental.

mitchdenny replied:
Oh its already in a preview package - never mind.

davidfowl replied:
Yea

captainsafia commented on tests/Aspire.Hosting.Docker.Tests/DockerComposePublisherTests.cs:
```diff
@@ -137,6 +137,8 @@ public async Task DockerComposeAppliesServiceCustomizations()
 
         builder.Services.AddSingleton<IResourceContainerImageBuilder, MockImageBuilder>();
 
+        var containerNameParam = builder.AddParameter("param1", "default-name", publishValueAsDefault: true);
```
```suggestion
        var containerNameParam = builder.AddParameter("param-1", "default-name", publishValueAsDefault: true);
```

To test the string replace logic.

captainsafia commented on src/Aspire.Hosting.Docker/DockerComposeServiceExtensions.cs:
```diff
@@ -48,4 +48,60 @@ public static IResourceBuilder<T> PublishAsDockerComposeService<T>(this IResourc
+    /// <summary>
+    /// Creates a placeholder for an environment variable in the Docker Compose file.
+    /// </summary>
+    /// <param name="parameter"></param>
```
Fill out these XML docs -- or ask Copilot to do it. 😛

captainsafia commented on src/Aspire.Hosting.Docker/DockerComposeServiceExtensions.cs:
```diff
@@ -48,4 +48,60 @@ public static IResourceBuilder<T> PublishAsDockerComposeService<T>(this IResourc
+    /// <param name="builder"></param>
+    /// <param name="dockerComposeService"></param>
+    /// <returns></returns>
+    public static string AsEnvPlaceHolder(this IResourceBuilder<ParameterResource> builder, DockerComposeServiceResource dockerComposeService)
```
This appears unused unless I missed something?

davidfowl replied:
This is just a new public api. Mirrors AsProvisioningParameter

captainsafia commented on src/Aspire.Hosting.Docker/DockerComposeServiceExtensions.cs:
```diff
@@ -48,4 +48,60 @@ public static IResourceBuilder<T> PublishAsDockerComposeService<T>(this IResourc
+        // Placeholder for resolving the actual parameter value
+        // https://docs.docker.com/compose/how-tos/environment-variables/variable-interpolation/#interpolation-syntax
+
+        // Treat secrets as environment variable placeholders as for now
```
Comment is irrelevant?

captainsafia commented on src/Aspire.Hosting.Docker/DockerComposeEnvironmentResource.cs:
```diff
@@ -62,4 +62,9 @@ private Task PublishAsync(PublishingContext context)
+
+    internal void AddEnvironmentVariable(string name, string? description = null, string? defaultValue = null)
+    {
+        CapturedEnvironmentVariables[name] = (description, defaultValue);
```
I don't think there's ever a case where we'll unintentionally override with duplicates. The only scenario I could think of would be if the parameter resource names were `parameter-1` and `parameter_1` and we don't permit underscores in resource names so I think this is fool-proof?

davidfowl replied:
There might be edge cases we need to deal with (a parameter named resource_image) or something crazy where the container image overrides the parameter env. I haven't thought through it much 😄

davidfowl committed:
PR feedback

davidfowl committed:
Enhance environment variable handling in Docker Compose resources

- Updated CapturedEnvironmentVariables to include a source object.
- Modified AddEnvironmentVariable method to accept a source parameter.
- Adjusted AsEnvironmentPlaceHolder methods to utilize the new source parameter for better context.

davidfowl committed:
Refactor AddEnvironmentVariable method to return environment variable placeholder and update usage in service extensions

davidfowl committed:
Add source parameter to AddEnvironmentVariable for container image reference

mitchdenny commented on src/Aspire.Hosting.Docker/DockerComposeServiceExtensions.cs:
```diff
@@ -48,4 +48,63 @@ public static IResourceBuilder<T> PublishAsDockerComposeService<T>(this IResourc
+    /// <param name="manifestExpressionProvider">The manifest expression provider.</param>
+    /// <param name="dockerComposeService">The Docker Compose service resource to associate the environment variable with.</param>
+    /// <returns>A string representing the environment variable placeholder in Docker Compose syntax (e.g., <c>${ENV_VAR}</c>).</returns>
+    public static string AsEnvironmentPlaceHolder(this IManifestExpressionProvider manifestExpressionProvider, DockerComposeServiceResource dockerComposeService)
```
Because I'm pedantic ... Placeholder not PlaceHolder

mitchdenny replied:
... but approving nonetheless.

davidfowl replied:
Fixed

davidfowl committed:
Rename AsEnvironmentPlaceHolder methods to AsEnvironmentPlaceholder for consistency

davidfowl committed:
Remove comment

davidfowl commented on src/Aspire.Hosting.Docker/DockerComposeServiceExtensions.cs:
```diff
@@ -48,4 +48,58 @@ public static IResourceBuilder<T> PublishAsDockerComposeService<T>(this IResourc
         return builder;
     }
+
+    /// <summary>
```
@prom3theu5 This is the pattern we want for helm values for the k8s manifest. I am thinking that before we GA either docker compose or the k8s resource types, the underlying model should be more typed so these aren't strings but something like `YamlValue<string>` (or maybe we go specific and say it's composevalue\<T\> and kubevalue\<T\>), that way.

We need to:
 - Allow going from app model to cdk model
 - Preserve the association between the cdk value and the aspire app model somewhere (for later)
